File: Chunk.py
from enum import IntEnum
from Object import *

class OpCode(IntEnum):
    OP_CONSTANT=0,
    OP_ADD=1,
    OP_SUB=2,
    OP_MUL=3,
    OP_DIV=4,
    OP_EQUAL=5,
    OP_GREATER=6,
    OP_LESS=7,
    OP_NOT=8,
    OP_MINUS=9,
    OP_AND=10,
    OP_OR=11,
    OP_JUMP_IF_FALSE=12,
    OP_JUMP=13,
    OP_SET_GLOBAL=14,
    OP_GET_GLOBAL=15,
    OP_SET_LOCAL=16,
    OP_GET_LOCAL=17,
    OP_ARRAY=18,
    OP_INDEX=19,
    OP_FUNCTION_CALL=20,
    OP_RETURN=21,
    OP_GET_BUILTIN_FUNCTION=22,
    OP_GET_BUILTIN_VARIABLE=23,
    OP_STRUCT=24,
    OP_GET_STRUCT=25,
    OP_SET_STRUCT=26,
    OP_REF_GLOBAL=27,
    OP_REF_LOCAL=28,
    OP_REF_INDEX_GLOBAL=29,
    OP_REF_INDEX_LOCAL=30,
    OP_SP_OFFSET=31,

class Chunk:
    opCodes:list[int]
    constants:list[Object]


    def __init__(self,opCodes:list[int],constants:list[Object]) -> None:
        self.constants=constants
        self.opCodes=opCodes

    def Stringify(self)->None:
        for i in range(0,len(self.constants)):
            if self.constants[i].type==ObjectType.FUNCTION:
                print("=======constant idx:%d    %s======="% (i,self.constants[i]))
                self.__OpCodeStringify(self.constants[i].opCodes)
                print()

        self.__OpCodeStringify(self.opCodes)

    def __OpCodeStringify(self,opcodes:list[int])->None:
        i = 0
        while i < (len(opcodes)):
            if opcodes[i]==OpCode.OP_CONSTANT:
                pos=opcodes[i+1]
                print("%8d    OP_CONSTANT    %d    '%s'" % (i,pos,self.constants[pos]))
                i=i+1
            elif opcodes[i]==OpCode.OP_ADD:
                print("%8d    OP_ADD" % (i))
            elif opcodes[i]==OpCode.OP_SUB:
                print("%8d    OP_SUB" % (i))
            elif opcodes[i]==OpCode.OP_MUL:
                print("%8d    OP_MUL" % (i))
            elif opcodes[i]==OpCode.OP_DIV:
                print("%8d    OP_DIV" % (i))
            elif opcodes[i]==OpCode.OP_LESS:
                print("%8d    OP_LESS" % (i))
            elif opcodes[i]==OpCode.OP_GREATER:
                print("%8d    OP_GREATER" % (i))
            elif opcodes[i]==OpCode.OP_MINUS:
                print("%8d    OP_MINUS" % (i))
            elif opcodes[i]==OpCode.OP_EQUAL:
                print("%8d    OP_EQUAL" % (i))
            elif opcodes[i]==OpCode.OP_ARRAY:
                count=opcodes[i+1]
                print("%8d    OP_ARRAY    %d" % (i,count))
                i=i+1
            elif opcodes[i]==OpCode.OP_INDEX:
                print("%8d    OP_INDEX" % (i))
            elif opcodes[i]==OpCode.OP_JUMP:
                address=opcodes[i+1]
                print("%8d    OP_JUMP    %d" % (i,address))
                i=i+1
            elif opcodes[i]==OpCode.OP_JUMP_IF_FALSE:
                address=opcodes[i+1]
                print("%8d    OP_JUMP_IF_FALSE    %d" % (i,address))
                i=i+1
            elif opcodes[i]==OpCode.OP_RETURN:
                count=opcodes[i+1]
                print("%8d    OP_RETURN    %d" % (i,count))
                i=i+1
            elif opcodes[i]==OpCode.OP_SET_GLOBAL:
                pos=opcodes[i+1]
                print("%8d    OP_SET_GLOBAL    %d" % (i,pos))
                i=i+1
            elif opcodes[i]==OpCode.OP_GET_GLOBAL:
                pos=opcodes[i+1]
                print("%8d    OP_GET_GLOBAL    %d" % (i,pos))
                i=i+1
            elif opcodes[i]==OpCode.OP_SET_LOCAL:
                isInUpScope=opcodes[i+1]
                scopeDepth=opcodes[i+2]
                index=opcodes[i+3]
                print("%8d    OP_SET_LOCAL    %d    %d    %d" % (i,isInUpScope,scopeDepth,index))
                i=i+3
            elif opcodes[i]==OpCode.OP_GET_LOCAL:
                isInUpScope=opcodes[i+1]
                scopeDepth=opcodes[i+2]
                index=opcodes[i+3]
                print("%8d    OP_GET_LOCAL    %d    %d    %d" % (i,isInUpScope,scopeDepth,index))
                i=i+3
            elif opcodes[i]==OpCode.OP_FUNCTION_CALL:
                argCount=opcodes[i+1]
                print("%8d    OP_FUNCTION_CALL    %d" % (i,argCount))
                i=i+1
            elif opcodes[i]==OpCode.OP_GET_BUILTIN_FUNCTION:
                builtinIdx=opcodes[i+1]
                print("%8d    OP_GET_BUILTIN_FUNCTION    %d" % (i,builtinIdx))
                i=i+1
            elif opcodes[i]==OpCode.OP_GET_BUILTIN_VARIABLE:
                builtinIdx=opcodes[i+1]
                print("%8d    OP_GET_BUILTIN_VARIABLE    %d" % (i,builtinIdx))
                i=i+1
            elif opcodes[i]==OpCode.OP_STRUCT:
                memberCount=opcodes[i+1]
                print("%8d    OP_STRUCT    %d" % (i,memberCount))
                i+=1
            elif opcodes[i]==OpCode.OP_GET_STRUCT:
                print("%8d    OP_GET_STRUCT" % (i))
            elif opcodes[i]==OpCode.OP_SET_STRUCT:
                print("%8d    OP_SET_STRUCT" % (i))
            elif opcodes[i]==OpCode.OP_REF_GLOBAL:
                idx=opcodes[i+1]
                print("%8d    OP_REF_GLOBAL    %d" % (i,idx))
                i=i+1
            elif opcodes[i]==OpCode.OP_REF_LOCAL:
                isInUpScope=opcodes[i+1]
                scopeDepth=opcodes[i+2]
                index=opcodes[i+3]
                print("%8d    OP_REF_LOCAL    %d    %d    %d" % (i,isInUpScope,scopeDepth,index))
                i=i+3
            elif opcodes[i]==OpCode.OP_REF_INDEX_GLOBAL:
                pos=opcodes[i+1]
                print("%8d    OP_REF_INDEX_GLOBAL    %d" % (i,pos))
                i=i+1
            elif opcodes[i]==OpCode.OP_REF_INDEX_LOCAL:
                isInUpScope=opcodes[i+1]
                scopeDepth=opcodes[i+2]
                index=opcodes[i+3]
                print("%8d    OP_REF_INDEX_LOCAL    %d    %d    %d" % (i,isInUpScope,scopeDepth,index))
                i=i+3
            elif opcodes[i]==OpCode.OP_SP_OFFSET:
                offset=opcodes[i+1]
                print("%8d    OP_SP_OFFSET    %d" % (i,offset))
                i=i+1
            i=i+1

File: Chunk.cpp
#include "Chunk.h"
#include "Object.h"

Chunk::Chunk(OpCodeList opCodeList, const std::vector<Value> &constants)
    : opCodeList(opCodeList), constants(constants)
{
}

std::string Chunk::Stringify()
{
    std::string result = OpCodeStringify(opCodeList);
    for (const auto &c : constants)
        if (IS_FUNCTION_VALUE(c))
            result += ObjectStringify(TO_FUNCTION_VALUE(c)
#ifndef NDEBUG
                                          ,
                                      true
#endif
            );
    return result;
}

std::string Chunk::OpCodeStringify(const OpCodeList &opCodeList)
{
    std::stringstream cout;
    for (int32_t i = 0; i < opCodeList.size(); ++i)
    {
        switch (opCodeList[i])
        {
        case OP_CONSTANT:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_CONSTANT\t'" << constants[opCodeList[++i]].Stringify() << "'" << std::endl;
            break;
        case OP_ADD:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_ADD" << std::endl;
            break;
        case OP_SUB:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_SUB" << std::endl;
            break;
        case OP_MUL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_MUL" << std::endl;
            break;
        case OP_DIV:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_DIV" << std::endl;
            break;
        case OP_LESS:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_LESS" << std::endl;
            break;
        case OP_GREATER:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_GREATER" << std::endl;
            break;
        case OP_NOT:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_NOT" << std::endl;
            break;
        case OP_MINUS:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_MINUS" << std::endl;
            break;
        case OP_EQUAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_EQUAL" << std::endl;
            break;
        case OP_ARRAY:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_ARRAY\t" << opCodeList[++i] << std::endl;
            break;
        case OP_AND:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_AND" << std::endl;
            break;
        case OP_OR:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_OR" << std::endl;
            break;
        case OP_BIT_AND:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_BIT_AND" << std::endl;
            break;
        case OP_BIT_OR:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_BIT_OR" << std::endl;
            break;
        case OP_BIT_NOT:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_BIT_NOT" << std::endl;
            break;
        case OP_BIT_XOR:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_BIT_XOR" << std::endl;
            break;
        case OP_GET_INDEX:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_GET_INDEX" << std::endl;
            break;
        case OP_SET_INDEX:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_SET_INDEX" << std::endl;
            break;
        case OP_JUMP:
#ifdef COMPUTEDUCK_BUILD_WITH_LLVM
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_JUMP\t" << opCodeList[++i] << "\t" << opCodeList[++i] << std::endl;
#else
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_JUMP\t" << opCodeList[++i] << std::endl;
#endif
            break;
        case OP_JUMP_IF_FALSE:
#ifdef COMPUTEDUCK_BUILD_WITH_LLVM
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_JUMP_IF_FALSE\t" << opCodeList[++i] << "\t" << opCodeList[++i] << std::endl;
#else
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_JUMP_IF_FALSE\t" << opCodeList[++i] << std::endl;
#endif
            break;
#ifdef COMPUTEDUCK_BUILD_WITH_LLVM
        case OP_JUMP_START:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_JUMP_START\t" << opCodeList[++i] << std::endl;
            break;
        case OP_JUMP_END:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_JUMP_END" << std::endl;
            break;
#endif
        case OP_RETURN:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_RETURN\t" << opCodeList[++i] << std::endl;
            break;
        case OP_DEF_GLOBAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_DEF_GLOBAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_SET_GLOBAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_SET_GLOBAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_GET_GLOBAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_GET_GLOBAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_DEF_LOCAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_DEF_LOCAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_SET_LOCAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_SET_LOCAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_GET_LOCAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_GET_LOCAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_GET_UPVALUE:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_GET_UPVALUE\t" << opCodeList[++i] << std::endl;
            break;
        case OP_SET_UPVALUE:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_SET_UPVALUE\t" << opCodeList[++i] << std::endl;
            break;
        case OP_CLOSURE:
        {
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_CLOSURE\t" << opCodeList[++i] << "\t" << opCodeList[++i] << std::endl;
            for (uint8_t j = 0; j < opCodeList[i]; ++j)
                cout << std::setfill('0') << std::setw(8) << i << "\t|\t" << opCodeList[++i] << "\t" << opCodeList[++i] << std::endl;
            break;
        }
        case OP_FUNCTION_CALL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_FUNCTION_CALL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_GET_BUILTIN:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_GET_BUILTIN\t'" << constants[opCodeList[++i]].Stringify() << "'" << std::endl;
            break;
        case OP_STRUCT:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_STRUCT\t" << opCodeList[++i] << std::endl;
            break;
        case OP_GET_STRUCT:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_GET_STRUCT" << std::endl;
            break;
        case OP_SET_STRUCT:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_SET_STRUCT" << std::endl;
            break;
        case OP_REF_GLOBAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_REF_GLOBAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_REF_LOCAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_REF_LOCAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_REF_INDEX_GLOBAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_REF_INDEX_GLOBAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_REF_INDEX_LOCAL:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_REF_INDEX_LOCAL\t" << opCodeList[++i] << std::endl;
            break;
        case OP_DLL_IMPORT:
            cout << std::setfill('0') << std::setw(8) << i << "\tOP_DLL_IMPORT" << std::endl;
            break;
        default:
            break;
        }
    }

    return cout.str();
}